Add mdn-bcd-collector update script

mzgoddard opened the pull request:
#### Summary

Add update script from https://github.com/GooborgStudios/mdn-bcd-collector. Add needed utilities (`logger` and `ua-parser`) and npm dependencies.

#### Test results and supporting details

I ran the update script locally with https://github.com/GooborgStudios/mdn-bcd-collector and https://github.com/GooborgStudios/mdn-bcd-results/tree/3af0b0ad1e7fd53098f266c80387e359288e21ad.

Running once completes without throwing an error.
Running a second time throws an error.

<details>
<summary>Error</summary>

```
~/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:120
        throw new Error(`Invalid argument not valid semver ('${version}' received)`);
              ^
Error: Invalid argument not valid semver ('13.1> 14.1' received)
    at validateAndParse (~/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:120:15)
    at compareVersions (~/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:10:16)
    at compareVersions (~/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:64:17)
    at update (file://~/browser-compat-data/scripts/update.ts:576:11)
    at main (file://~/browser-compat-data/scripts/update.ts:686:22)
    at async file://~/browser-compat-data/scripts/update.ts:746:3
```

</details>

I made a number of changes to the script:

- Add constants from `mdn-bcd-collector/lib/constants.ts` to `update` script
- Add relevant types from `mdn-bcd-collector/types/types.ts` to `update` script
- Replace `Array<string, string, string, TestResultValue>` in `Overrides` with `[string, string, string, TestResultValue]`
- Add type definition `ManualOverride = [string, string, string, TestResultValue]`
- Type `Array.isArray` in `overrides.filter` as `(item: unknown) => item is ManualOverride`
- Add `MDN_COLLECTOR_DIR` for loading overrides file
- Changed overrides file path to `path.join(MDN_COLLECTOR_DIR, 'custom/overrides.json'),`

Dependencies added to package:

- `fs-extra`
- `klaw`
- `ua-parser-js`
- `winston`
- `@google-cloud/logging-winston`

#### Related issues

https://github.com/mdn/browser-compat-data/issues/19868
https://github.com/mdn/browser-compat-data/issues/18514

mzgoddard commented:
After speaking with @foolip I've made this as a first step in reducing tech debt in the update script by moving it into browser-compat-data. I tried to keep the number of changes low to the script itself low.

There will be a lot of steps to do after merging this PR that we can list in #19868 or its own issue.

foolip commented on package.json:
```diff
@@ -30,6 +30,7 @@
     "@babel/eslint-parser": "~7.21.3",
     "@babel/plugin-syntax-import-assertions": "~7.20.0",
     "@desertnet/html-parser": "~1.0.1",
+    "@google-cloud/logging-winston": "5.3.0",
```
This dependency in mdn-bcd-collector was really there to make logging in GCP work. For a script run locally, here in BCD, I think we should use the logging from test/utils.ts.

foolip commented on package.json:
```diff
@@ -61,16 +62,20 @@
     "typescript": "~5.0.2",
+    "ua-parser-js": "1.0.35",
     "web-specs": "^2.41.0",
+    "winston": "3.8.2",
```
Let's see if we can avoid this dependency.

mzgoddard replied:
I replaced the `logger` file with a small shim I added to `update`.

foolip commented on package.json:
```diff
@@ -61,16 +62,20 @@
     "eslint-plugin-unicorn": "^47.0.0",
     "fast-json-stable-stringify": "~2.1.0",
     "fdir": "~6.0.1",
+    "fs-extra": "11.1.1",
```
I think this is only used for `fs.readJson()`. Can you get rid of it and just use plain fs and `JSON.parse()` like other existing code in BCD?

mzgoddard replied:
As part of this I'm going to make reading the json files serial like what is done in `index.ts`. `update` at the moment reads the json in parallel. Since it is as of this time 1455 json files, some systems may run in EMFILE or ENFILE errors opening that many files at one time.

foolip replied:
Makes sense, I guess @queengooborg and I never ran into that.

foolip replied:
@mzgoddard the fs-extra dependency is still here. Can it be removed now?

foolip commented on package.json:
```diff
@@ -61,16 +62,20 @@
+    "fs-extra": "11.1.1",
     "husky": "^8.0.1",
     "json-schema-to-typescript": "~10.1.5",
+    "klaw": "4.1.0",
```
Can you use fdir instead? That's an existing dependency.

mzgoddard replied:
Its now using fdir with `withFullPaths`.

foolip commented on utils/logger.ts:
```diff
@@ -0,0 +1,32 @@
+// See the LICENSE file for copyright details
+//
+
+import winston from 'winston';
```
I hope this won't be needed, console.log and chalk FTW!

foolip commented on scripts/update.ts:
```diff
@@ -0,0 +1,748 @@
+//
+// mdn-bcd-collector: scripts/update-bcd.ts
```
Can you use the same header as elsewhere in BCD?

I approve using the code under the BCD license, which isn't the same as in mdn-bcd-collector. @queengooborg can you also approve this?

mzgoddard replied:
I replaced the headers with the common one in BCD.

foolip commented:
@mzgoddard this looks good to me now. Can you try running the script with various inputs, and see if makes exactly the same local changes to BCD, as running it from @queengooborg's repo would?

mzgoddard commented:
Including `@types/minimatch` in our dependencies doesn't fix it if we use a version of `minimatch` that includes types. `@types/glob` imports `minimatch`, when typescript is type checking, it looks at `minimatch` first, if `minimatch` doesn't have types it searches for `@types/minimatch`.

I've added a commit that rolls `minimatch` to version `5`. It includes `@types/minimatch@5`. This is able to install and tests are able to run and pass. I did have to change how it is imported, version `5` doesn't export ecmascript modules and so importing `{ Minimatch }` is separate from the import. I'm not what impacts this may have for filtering identifier paths in the update script.

mzgoddard commented:
I still need to test running the script again and seeing how it meshes with the current data.

mzgoddard commented:
I got the following results running with `mdn-bcd-collector` and `mdn-bcd-results` checked out to https://github.com/GooborgStudios/mdn-bcd-collector/commit/503072afb8de3669972c2aae82a9b6f928f9b26d and https://github.com/GooborgStudios/mdn-bcd-results/commit/6c807fd253581cb1b417bd2b26af8239b1a1ac1d. The mdn-bcd-collector version is ~2 weeks ago but is the version I pulled the update script from.

Running the script results with these changes:

<details>
<summary><code>$ git status</code></summary>

```
$ git status
On branch update-script
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   api/DeprecationReportBody.json
        modified:   api/Document.json
        modified:   api/DragEvent.json
        modified:   api/EXT_color_buffer_float.json
        modified:   api/Element.json
        modified:   api/HTMLCanvasElement.json
        modified:   api/InstallEvent.json
        modified:   api/InterventionReportBody.json
        modified:   api/MediaQueryList.json
        modified:   api/MediaStreamTrackProcessor.json
        modified:   api/NetworkInformation.json
        modified:   api/OES_fbo_render_mipmap.json
        modified:   api/OVR_multiview2.json
        modified:   api/OffscreenCanvasRenderingContext2D.json
        modified:   api/PannerNode.json
        modified:   api/PictureInPictureEvent.json
        modified:   api/Plugin.json
        modified:   api/PushManager.json
        modified:   api/PushSubscription.json
        modified:   api/RTCPeerConnection.json
        modified:   api/RTCRtpSender.json
        modified:   api/RTCStatsReport.json
        modified:   api/ReadableStreamDefaultController.json
        modified:   api/ReadableStreamDefaultReader.json
        modified:   api/Report.json
        modified:   api/ReportBody.json
        modified:   api/Request.json
        modified:   api/SVGImageElement.json
        modified:   api/ScreenOrientation.json
        modified:   api/ServiceWorkerRegistration.json
        modified:   api/SpeechRecognitionAlternative.json
        modified:   api/SpeechRecognitionErrorEvent.json
        modified:   api/SpeechRecognitionEvent.json
        modified:   api/SpeechRecognitionResult.json
        modified:   api/SpeechRecognitionResultList.json
        modified:   api/SpeechSynthesisErrorEvent.json
        modified:   api/SpeechSynthesisEvent.json
        modified:   api/SpeechSynthesisVoice.json
        modified:   api/Touch.json
        modified:   api/TouchList.json
        modified:   api/TransformStreamDefaultController.json
        modified:   api/USBPermissionResult.json
        modified:   api/VRPose.json
        modified:   api/VideoPlaybackQuality.json
        modified:   api/WEBGL_compressed_texture_astc.json
        modified:   api/WEBGL_compressed_texture_etc.json
        modified:   api/WEBGL_compressed_texture_pvrtc.json
        modified:   api/WebGL2RenderingContext.json
        modified:   api/WebGLQuery.json
        modified:   api/WebGLRenderingContext.json
        modified:   api/WebGLSampler.json
        modified:   api/WebGLSync.json
        modified:   api/XRFrame.json
        modified:   api/XRHand.json
        modified:   api/XRInputSource.json
        modified:   api/XRJointPose.json
        modified:   api/XRJointSpace.json
        modified:   api/XRSystem.json
        modified:   css/properties/appearance.json
        modified:   css/properties/aspect-ratio.json
        modified:   css/properties/background-clip.json
        modified:   css/properties/background-image.json
        modified:   css/properties/caption-side.json
        modified:   css/properties/contain-intrinsic-height.json
        modified:   css/properties/contain-intrinsic-width.json
        modified:   css/properties/font-variant-emoji.json
        modified:   css/properties/image-rendering.json
        modified:   css/properties/line-height.json
        modified:   css/properties/list-style-type.json
        modified:   css/properties/mask-clip.json
        modified:   css/properties/mask-image.json
        modified:   css/properties/overflow-clip-box-block.json
        modified:   css/properties/overflow-clip-box-inline.json
        modified:   css/properties/overflow-clip-box.json
        modified:   css/properties/page.json
        modified:   css/properties/scroll-snap-type-x.json
        modified:   css/properties/tab-size.json
        modified:   css/properties/text-align.json
        modified:   css/properties/text-emphasis-position.json
        modified:   css/properties/text-orientation.json
        modified:   css/properties/text-overflow.json
        modified:   css/properties/text-underline-position.json
        modified:   css/properties/transition.json
        modified:   css/properties/user-modify.json
        modified:   css/properties/user-select.json
        modified:   css/properties/width.json
        modified:   javascript/builtins/WebAssembly/Module.json

no changes added to commit (use "git add" and/or "git commit -a")
```

</details>

Running a second time produces the following error:

```
warn: api.Document.writeln skipped for safari due to multiple default statements
warn: api.Document.xmlEncoding skipped for firefox due to added+removed statement
warn: api.Document.xmlStandalone skipped for firefox due to added+removed statement
warn: api.Document.xmlVersion skipped for firefox due to added+removed statement
/Users/zen/Code/bocoup/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:120
        throw new Error(`Invalid argument not valid semver ('${version}' received)`);
              ^
Error: Invalid argument not valid semver ('13.1> 14.1' received)
    at validateAndParse (/Users/zen/Code/bocoup/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:120:15)
    at compareVersions (/Users/zen/Code/bocoup/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:10:16)
    at compareVersions (/Users/zen/Code/bocoup/browser-compat-data/node_modules/compare-versions/lib/esm/index.js:64:17)
    at update (file:///Users/zen/Code/bocoup/browser-compat-data/scripts/update.ts:580:11)
    at main (file:///Users/zen/Code/bocoup/browser-compat-data/scripts/update.ts:693:22)
    at async file:///Users/zen/Code/bocoup/browser-compat-data/scripts/update.ts:753:3
```

This semver is stored as `"13.1> ≤14.1"`. Before the version is compared `≤` is replaced by an empty string `''`.

I think the script is not operating as expected. Here are some snippets from a diff:

```diff
-                "version_added": "113"
+                "version_added": false

-              "version_added": "3.1"
+              "version_added": "13.1> ≤14.1"

-            "version_added": "51"
+            "version_added": "80"

-                "version_added": "79",
-                "flags": [
-                  {
-                    "type": "preference",
-                    "name": "#enable-experimental-web-platform-features"
-                  }
-                ]
+                "version_added": "≤80"

-              "edge": "mirror",
+              "edge": {
+                "version_added": "≤80"
+              },
```

Should those changes be made by the update script?

foolip commented:
@mzgoddard that you get a "Invalid argument not valid semver ('13.1> 14.1' received)" error the second time is expected, since the script adds '13.1> 14.1' and that's invalid. This should be the same for the original script too.

The comparison I'd like to see is if there are any differences in the changes made when the original script is run once, vs. when the relocated script in this PR is run once. The way I'd compare it is by creating a temporary commit for each and using `git diff`. If there are any differences at all, then moving the script must have accidentally introduced changes.

queengooborg commented:
In the BCD meeting, I had asked if we should be importing other collector resources and interacting with the collector repository on a scripting level, as so far we have only been linking to it in documentation.  The consensus is to go for it, especially as we're already so dependent on the collector!

As our current primary goal for BCD is to fix discrepancies between the collector and BCD for browsers released in 2020+ (https://github.com/openwebdocs/project/issues/85), I'm afraid I won't be able to provide too much review assistance on this PR, as the goal is to complete that project by the end of the quarter.  I will try to review what I can though!

foolip commented:
@mzgoddard can you resolve conflicts and then compare the changes made by this script vs. the original?

mzgoddard commented:
I resolved conflicts and tested this script vs the one in GooborgStudios/mdn-bcd-collector.

There were no differences between the changes both scripts made.

Here is a gist with the diffs of each: https://gist.github.com/mzgoddard/b1ff3687630db9c6cf05579fb8413674

foolip commented:
Brilliant, thank you @mzgoddard! I'll go ahead and merge this before there are more conflicts.

foolip commented:
Oh no, the Node LTS CI job is failing. @mzgoddard can you check if it's caused by these changes?

mzgoddard commented:
@foolip I checked out the CI job. Its an error from `swc` during `generate-types`.

> `@swc/core threw an error when attempting to validate swc compiler options.`

I don't think its related to the change in the PR. I think its a flaky error. If you re-run the job I think it will pass. (I don't have the permission needed to re-run.)

foolip commented:
@mzgoddard I re-ran the job but it's still failing. Unless it's also failing in other recent PRs I guess it was caused by the changes here?

queengooborg commented:
To help reduce potential for merge conflicts, we may want to split this down into even smaller chunks.  For example, the UA parser I think could be its own PR (and I have some change requests regarding it specifically that would help turn it into a useful utility).

mzgoddard commented:
I found the cause of the issue in CI testing. I made a mistake on one of the `package-lock.json` merges and it was persisting an older `@swc/core` dependency. I've checked out `package-lock.json` from `main` and reinstalled the new dependencies to resolve this.

mzgoddard commented:
> To help reduce potential for merge conflicts, we may want to split this down into even smaller chunks. For example, the UA parser I think could be its own PR (and I have some change requests regarding it specifically that would help turn it into a useful utility).

I'll make a PR with just UA parser and its test file. (That may not reduce merge conflicts by much since UA parser has its own dependency, `ua-parser-js` and the merge conflicts are happening in `package.json` and `package-lock.json`.)

foolip reviewed (approved):
Thank you @mzgoddard for sorting out the CI issues and sending https://github.com/mdn/browser-compat-data/pull/20210 separately. I'll go ahead and merge this now.
